StrongInject.Generator/Visitors/DependencyCheckerVisitor.cs

```diff
@@ -249,17 +252,15 @@ public bool IsScopeAsync
             public HashSet<IParameterSymbol>? UsedParams { get; set; }
 
 
-            public bool AddCurrentlyVisiting(InstanceSource source) =>
-                source.Scope == Scope.SingleInstance
-                    ? CurrentlyVisitingSingletons.Add(source)
-                    : CurrentlyVisitingNonSingletons.Add(source);
-            
-            public bool RemoveCurrentlyVisiting(InstanceSource source) =>
-                source.Scope == Scope.SingleInstance
-                    ? CurrentlyVisitingSingletons.Remove(source)
-                    : CurrentlyVisitingNonSingletons.Remove(source);            
-            public HashSet<InstanceSource> CurrentlyVisitingSingletons { get; set; }
-            public HashSet<InstanceSource> CurrentlyVisitingNonSingletons { get; set; }
+            public bool AddCurrentlyVisiting(InstanceSource source, InstanceSourcesScope scope, out InstanceSourcesScope existingScope)
+            {
+                _currentlyVisiting.AddOrUpdate(source, scope, out existingScope);
+                return existingScope is null;
+            }
+
+            public bool RemoveCurrentlyVisiting(InstanceSource source) => _currentlyVisiting.Remove(source);
+
+            public void ResetCurrentlyVisiting() => _currentlyVisiting = new();
```

Comment:
Would `_currentlyVisiting.Clear()` potentially be desirable?

Reply:
nope - we need to preserve this for continuing higher up the stack.